### DISCUSSION TITLE
Running in webbrowsers

### DISCUSSION THREAD

**Discussion by jschwartzenberg** [dosemu2/dosemu2#2213] [Ideas]
I just came across: https://emupedia.net/beta/emuos/

It seems most of this stuff runs full-speed on my old Intel Core2 Duo laptop from 2007/2008 with Firefox. Quite impressive. Might it be possible to run DOSEMU2 like that too?

**Comment by stsp** [comment]
This means compiling dosemu2 with
emscripten. It seems, these days
emscripten is very advanced, and
offers the shared mmap() support,
so that dosemu2 can do its MMU
tricks there perhaps.
#575 lists bios.S linking as a problem,
but since that time IIRC Andrew did
a really great job at solving that one
(yes, there were the days when Andrew
did the core work, not only CI :)
So currently I don't know what keeps
us from the working emscripten port.
I _suppose_ there are 1024+ hard spots,
but none are currently known.
Also dosemu2 suite now consists of
dosemu2+fdpp+comcom64. So porting
dosemu2 alone is no longer enough.
Well, fdpp is small, but comcom64+dj64dev
is not so small. This all needs to be
studied.

**Comment by stsp** [comment]
I added the emscripten support now.
It compiles well.
I wrote the list of missing functions
in #575 
They are quite a few, and all but the
ones from ucontext.h are easy to
replace. But ucontext probably can't
be hand-crafted, and I don't think
dosemu2 can easily run w/o those,
as it is now fully coopthreadized.
dosemu1 didn't require those, but
in dosemu2 they are not optional...

**Reply by jschwartzenberg** [reply]
Great to see progress there! It seems DOSEMU2 is becoming quite portable, I should look at testing more again :)

**Reply by stsp** [reply]
Yes, its quite portable, but emscripten
supports only a small subset of posix
functionality. So while most functions
are there, some of them do nothing,
and some - ignore particular flag combinations.
I am very sceptical that things will work
as expected. Because even on linux they
usually work as expected only after I
submit patches or bug reports.
